dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java

```diff
@@ -2288,62 +2290,69 @@ final protected void sendHeaders(HttpBaseMessageImpl msg) throws IOException {
+            if (acceptEncoding != null) {
+                ResponseCompressionHandler compressionHandler = new ResponseCompressionHandler(getHttpConfig(), nettyResponse, acceptEncoding);
+                compressionHandler.setCurrentContentLength(getResponse().getContentLength());
+                compressionHandler.process();
+                if (compressionHandler.getEncoding() != null) {
+                    setupCompressionHandler(compressionHandler.getEncoding());
+                    ContentEncodingValues ce = this.compressHandler.getContentEncoding();
+                    getResponse().setContentEncoding(ce);
                 }
-                response.headers().set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
             }
-            if (HttpUtil.isContentLengthSet(response)) {
-                this.nettyContext.channel().attr(NettyHttpConstants.CONTENT_LENGTH).set(HttpUtil.getContentLength(response));
+        }
+        ((NettyResponseMessage) getResponse()).processCookies();
+        HeaderHandler headerHandler = new HeaderHandler(myChannelConfig, response);
+        headerHandler.complianceCheck();
+        String closeNonUpgraded = (String) (this.myVC.getStateMap().get(TransportConstants.CLOSE_NON_UPGRADED_STREAMS));
+        // Shouldn't close upgraded requests
+        boolean upgradedRequest = closeNonUpgraded != null && closeNonUpgraded.equalsIgnoreCase("true");
+        if (!upgradedRequest && (!myChannelConfig.isKeepAliveEnabled() || (myChannelConfig.getMaximumPersistentRequests() != -1 && nettyContext.channel().attr(NettyHttpConstants.NUMBER_OF_HTTP_REQUESTS).get() >= myChannelConfig.getMaximumPersistentRequests()))) {
+            // Keep alive disabled or exceeded maximum number of keep alive requests
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "sendHeaders: Adding close connection header due to keep alive disabled or exceeded number of maximum persistent requests");
             }
-
+            response.headers().set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
+        }
+        if (HttpUtil.isContentLengthSet(response)) {
+            this.nettyContext.channel().attr(NettyHttpConstants.CONTENT_LENGTH).set(HttpUtil.getContentLength(response));
         }
         final boolean isSwitching = response.status().equals(HttpResponseStatus.SWITCHING_PROTOCOLS);
 
         if (isSwitching && "websocket".equalsIgnoreCase(response.headers().get(HttpHeaderNames.UPGRADE))) {
             nettyContext.channel().attr(NettyHttpConstants.PROTOCOL).set("WebSocket");
         }
-        this.nettyContext.channel().writeAndFlush(response);
-
         this.setHeadersSent();
+    }
+
+    final protected void sendHeaders(HttpResponse response) throws IOException {
+        if (headersSent()) {
+            Tr.event(tc, "Invalid call to sendHeaders after already sent");
```

Comment:
Do we want an event here? If so it should be added to the translations list so we don't forget about it. I assume this comes out in the messages/console.log?

Reply:
This is the same message and tracing used in legacy https://github.com/OpenLiberty/open-liberty/blob/c8249ca16f2cb8e01cf5b098402ebedd7e7817f7/dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java#L2192-L2195